### PR TITLE
[@ember/routing] add RouteInfoWithAttributes, etc,

### PR DESCRIPTION
**Pull request by jamescdavis**
This add [RouteInfoWithAttributes](https://api.emberjs.com/ember/3.6/classes/RouteInfoWithAttributes) and uses it in appropriate places. It also adds [buildRouteInfoMetadata](https://api.emberjs.com/ember/3.18/classes/Route/methods/buildRouteInfoMetadata?anchor=buildRouteInfoMetadata) to `@ember/routing/route` and [recognize](https://api.emberjs.com/ember/3.6/classes/RouterService/methods/recognize?anchor=recognize), [recognizeAndLoad](https://api.emberjs.com/ember/3.6/classes/RouterService/methods/recognize?anchor=recognizeAndLoad), and [rootURL](https://api.emberjs.com/ember/2.16/classes/RouterService/properties/rootURL?anchor=rootURL) to `@ember/routing/router-service`.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://api.emberjs.com/ember/3.6/classes/RouteInfoWithAttributes
- [ ] ~~If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.~~
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] ~~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.~~

### CODE REVIEW

**Comment by jamescdavis** [comment]
@mike-north @chriskrycho @dfreeman not sure where the bot is. vOv

**Review by chriskrycho — APPROVED** [review]

File: types/ember__routing/router-service.d.ts [hunk 1]
```diff
@@ -265,4 +266,57 @@ export default class RouterService extends Service {
...
+     *     modulePrefix: 'router-service',
+     *     environment,
+     *     rootURL: '/my-root',
+     *   …
+     *   }
+     * ]
+     * ```
+     * This property will return `/my-root`.
+     */
+    rootURL: string;
```
Comment (chriskrycho) [inline types/ember__routing/router-service.d.ts:321]: I think this should be `readonly`?

**Review by chriskrycho — CHANGES_REQUESTED** [review]
Caught one thing after the initial pass. Thanks for adding this!

**Comment by jamescdavis** [comment]
@chriskrycho updated

**Review by chriskrycho — CHANGES_REQUESTED** [review]
With apologies for another Request Changes—I should have thought of this when making my suggestion earlier—but we should `// $ExpectType Readonly<string>` etc. for all the `readonly` types.

**Comment by jamescdavis** [comment]
@chriskrycho 🤔  wait, I don't think `Readonly` works that way: https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype

**Comment by chriskrycho** [comment]
Ah, I believe you're correct, it's actually the `readonly` modifier proper, I believe, like this:

```ts
router.rootURL; // $ExpectType readonly string
```

That said, I haven't poked at that corner of DTSLint in a *while*, so I'm not positive that works outside the context of arrays. 🤔

**Comment by jamescdavis** [comment]
@chriskrycho indeed, `readonly` only works for arrays and tuples marked as `readonly`. AFAICT, there's no way to check that a property is marked as `readonly`. To try to find an example, I ran a case-insensitive grep of all tests in DT for `$ExpectType.*readonly` and the handful of matches are all for arrays or tuples.

**Comment by jamescdavis** [comment]
Oh, wait, 🤦 :
```ts
router.rootURL = 'foo'; // $ExpectError
```

**Comment by jamescdavis** [comment]
@orta, as another data point, @typescript-bot seems to be ignoring this one too (I pushed a revision and it hasn't updated status, re-requested review, etc).

**Comment by chriskrycho** [comment]
Hahaha, that’s a good call. Should have thought of that myself.

File: types/ember__routing/-private/route-info-with-attributes.d.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * It is read-only and internally immutable. It is also not
+ * observable, because a Transition instance is never
+ * changed after creation.
+ */
+export default interface RouteInfoWithAttributes extends RouteInfo {
+    /**
+     * This is the resolved return value from the
+     * route's model hook.
+     */
+    readonly attributes: any;
```
Comment (chriskrycho) [inline types/ember__routing/-private/route-info-with-attributes.d.ts:18]:
```suggestion
    readonly attributes: unknown;
```

File: types/ember__routing/-private/route-info-with-attributes.d.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    /**
+     * Will contain the result `Route#buildRouteInfoMetadata`
+     * for the corresponding Route.
+     */
+    readonly metadata: any;
```
Comment (chriskrycho) [inline types/ember__routing/-private/route-info-with-attributes.d.ts:23]:
```suggestion
    readonly metadata: unknown;
```

File: types/ember__routing/route.d.ts [hunk 4]
```diff
@@ -493,4 +493,34 @@ export default class Route extends EmberObject.extend(ActionHandler, Evented) {
...
+     *     super(...arguments);
+     *     this.router.on('routeDidChange', transition => {
+     *       document.title = transition.to.metadata.title;
+     *       // would update document's title to "Posts Page"
+     *     });
+     *   }
+     * }
+     * ```
+     */
+    buildRouteInfoMetadata(): any;
```
Comment (chriskrycho) [inline types/ember__routing/route.d.ts:525]:
```suggestion
    buildRouteInfoMetadata(): unknown;
```

**Review by chriskrycho — CHANGES_REQUESTED** [review]
Ugh, sorry, one more round.

**Comment by jamescdavis** [comment]
@chriskrycho good call on `unknown`. I guess I was translating a little too literally from the docs! 😆

**Review by chriskrycho — APPROVED** [review]
:shipit:

**Comment by chriskrycho** [comment]
Ready to merge

**Comment by chancancode** [comment]
eh, should the bot have required the files to be up-to-date? https://github.com/DefinitelyTyped/DefinitelyTyped/pull/49340 was merged earlier today and now there is a soft-conflict (the `rootURL` property is declared twice).

**Comment by chancancode** [comment]
#49345 will resolve the conflict, but seems like something that the bot should account for/enforce

**Pull request merged**